### ISSUE TITLE
"include old metricdefs up to 24h" prevents from new higher res data to become visible

### ISSUE THREAD

**Issue by Dieterbe** [grafana/metrictank#380]
if you switch over from sending a series as low-res to high-res, the index will incorporate metricdefs up to 24h prior to the `from` of the query, and because we incorporate those old defs when we shouldn't, we force other series in the same request to be of the same resolution, so we sample them down.
this means you have to wait 24 for high res to become visible as high res.

**Comment by Dieterbe** [comment]
we still have: 
```
func (s *Server) indexFind(ctx *middleware.Context, req models.IndexFind) {
	// metricDefs only get updated periodically (when using CassandraIdx), so we add a 1day (86400seconds) buffer when
	// filtering by our From timestamp.  This should be moved to a configuration option
	if req.From != 0 {
		req.From -= 86400
	}
```

**Comment by woodsaj** [comment]
This stray req.From adjustment is causing issues for some customers, namely it leads to alerts flapping.
When handling a query, if the series exists on the local node it will be filtered from results when it has not been updated since the query From time.  But if the series exists on a peer node it will only be filtered from the results when it has not been updated for 1day before the from timestamp.

I dont think we should just use the From time as the lastSeen ts for the index search.  This causes problems for alerting queries in the form
```
A=transformNull(service.*.uptime, 0)
query (A, 20m, now)
```
this type of query is just making sure the uptime metric for all services have seen at least 1 non-zero value in the last 20minutes.   However if we use the From (now-20m) as the lastSeen value on index searches the alert probably won't fire for services that go offline and don't send any metrics.   When the service has been offline for 19minutes, all datapoints except the oldest will be null.  But at 20mins the series will be filtered as it has not been updated on or after the From timestamp.

So i think we need to make lastSeen to be older then From, but by how much.  For some use cases 1day is practical, in others it is not.
The options that I see are
1) we just use 1day and document it
2) we use a configurable duration that can be customized on a per deployment basis
3) We use a configurable number of intervals instead of a fixed duration.  eg, the index should return all series that have been updated on or since `lastSeen-(def.Interval * padding)`

**Comment by replay** [comment]
I think this should definitely be configurable to be able to accommodate different sparsity and intervals, so option 1 is out.
Option 3 seems kind of intuitive but it also makes reasoning about the behavior a little more complicated when somebody uses a large number of different intervals, and i'm not sure if there is a real benefit to it.
So I'd choose option 2, as it's simple and should be sufficient to solve the issue.

**Comment by Dieterbe** [comment]
remember that graphite/whisper does not do any such filtering afaik. but we do, on the premise that otherwise queries would get too expensive.  we could also just remove the filtering by default, but make it available via an api function. and if people run into query limits we can suggest using that function.

> I think this should definitely be configurable to be able to accommodate different sparsity and intervals, so option 1 is out.

can you give an example of where a certain sparsity or interval becomes problematic with a hardcoded 1day value. i suspect these cases are so rare that the 24h tradeoff might still be quite good.

we can also consider using a different mechanism in the two use cases (in my originally reported use case, we can more stricter filter out series if we know we have a higher-res replacement for it)

**Comment by Dieterbe** [comment]
@replay see question above, thanks :)

**Comment by replay** [comment]
@Dieterbe f.e. in our backups we have an interval of 24h for some of the backup metrics. Those are not always collected at the same second of the day, sometimes they might be a few min later and sometimes a few min earlier (depending on how long the backup took). So if the limit is hard-coded to `24h` we might see certain series at one time and not see them at another.
Generally I think there's no disadvantage to making that value configurable, so even if we can't think of an edge-case that requires this to be changed it might still be worth it to just add a parameter for it.

**Comment by Dieterbe** [comment]
fixed by #767 except for the new case brought up here which now has ticket #838
